File: apps/api/models/vault.py
"""Vault models: company_credentials (encrypted fields)."""

import uuid
from typing import Optional

from sqlalchemy import ForeignKey, String
from sqlalchemy.dialects.postgresql import ARRAY, UUID
from sqlalchemy.orm import Mapped, mapped_column

from packages.common.db.base import Base, TimestampMixin, UUIDMixin


class CompanyCredential(Base, UUIDMixin, TimestampMixin):
    __tablename__ = "company_credentials"

    created_by: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=False
    )
    last_modified_by: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("profiles.id"), nullable=True
    )
    linked_product_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("products.id"), nullable=True
    )
    label: Mapped[str] = mapped_column(String, nullable=False)
    category: Mapped[str] = mapped_column(String, default="general")
    service_name: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    url: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    username_encrypted: Mapped[Optional[str]] = mapped_column(
        String, nullable=True
    )
    password_encrypted: Mapped[Optional[str]] = mapped_column(
        String, nullable=True
    )
    email_encrypted: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    api_secret_encrypted: Mapped[Optional[str]] = mapped_column(
        String, nullable=True
    )
    notes_encrypted: Mapped[Optional[str]] = mapped_column(String, nullable=True)
    tags: Mapped[Optional[list[str]]] = mapped_column(
        ARRAY(String), nullable=True
    )

File: apps/web/src/lib/types/vault.ts
export type CredentialCategory =
  | "api_key"
  | "service_login"
  | "database"
  | "cloud_provider"
  | "domain_registrar"
  | "email_service"
  | "payment_gateway"
  | "monitoring"
  | "other";

export interface CompanyCredential {
  id: string;
  label: string;
  category: string;
  service_name: string | null;
  url: string | null;
  username_encrypted: string | null;
  email_encrypted: string | null;
  password_encrypted: string | null;
  api_secret_encrypted: string | null;
  notes_encrypted: string | null;
  tags: string[] | null;
  linked_product_id: string | null;
  created_by: string;
  last_modified_by: string | null;
  created_at: string;
  updated_at: string;
}
